Sources/GRPC/Interceptor/ClientTransport.swift

```diff
@@ -0,0 +1,752 @@
+/// interceptor pipeline – are emitted to the `onResponsePart` callback supplied to the initializer.
+///
+/// In most instances the glue code is simple: transformations are applied to the request and
+/// response types used by the interceptor pipeline and the `NIO.Channel`. In addition, the
+/// transport keeps track of the state of the call and the `Channel`, taking appropriate action
+/// when these change. This includes buffering request parts from the interceptor pipeline until
+/// the `NIO.Channel` becomes active.
+///
+/// ### Thread Safety
+///
+/// This class is not thread safe. All methods **must** be executed on the transport's `eventLoop`.
+internal final class ClientTransport<Request, Response> {
+  /// The `EventLoop` this transport is running on.
+  private let eventLoop: EventLoop
+
+  /// The current state of the transport.
+  private var state: State = .idle
+
+  // Note: initial capacity is 4 because it's a power of 2 and most calls are unary so will
+  // have 3 parts.
+  /// A buffer to store request parts and promises in before the channel has become active.
+  private var writeBuffer = MarkedCircularBuffer<RequestAndPromise>(initialCapacity: 4)
+
+  /// A request part and a promise.
+  private struct RequestAndPromise {
+    var request: ClientRequestPart<Request>
+    var promise: EventLoopPromise<Void>?
+  }
+
+  /// Details about the call.
+  private let callDetails: CallDetails
+
+  /// A logger.
+  private var logger: Logger {
+    return self.callDetails.options.logger
+  }
+
+  /// Is the call streaming requests?
+  private var isStreamingRequests: Bool {
+    switch self.callDetails.type {
+    case .unary, .serverStreaming:
+      return false
+    case .clientStreaming, .bidirectionalStreaming:
+      return true
+    }
+  }
+
+  // Our `NIO.Channel` will fire trailers and the `GRPCStatus` to us separately. It's more
+  // convenient to have both at the same time when intercepting response parts. We'll hold on to the
+  // trailers here and only forward them when we receive the status.
+  private var trailers: HPACKHeaders?
+
+  /// The interceptor pipeline connected to this transport. This must be a `var` and an IOU because
+  /// it requires `self` to be
+  private var pipeline: ClientInterceptorPipeline<Request, Response>!
+
+  /// Our current state as logging metadata.
+  private var stateForLogging: Logger.MetadataValue {
+    if self.state.mayBuffer {
+      return "\(self.state) (\(self.writeBuffer.count) parts buffered)"
+    } else {
+      return "\(self.state)"
+    }
+  }
+
+  internal init(
+    details: CallDetails,
+    eventLoop: EventLoop,
+    interceptors: [ClientInterceptor<Request, Response>],
+    errorDelegate: ClientErrorDelegate?,
+    _ onResponsePart: @escaping (ClientResponsePart<Response>) -> Void
+  ) {
+    self.eventLoop = eventLoop
+    self.callDetails = details
+    self.pipeline = ClientInterceptorPipeline(
```

Comment:
Do we need to `nil` this out on `handlerRemoved` to break the reference cycle?

Reply:
Yes, we should. Good catch!